### PR TITLE
changefeedccl: support bit and varbit column types in avro changefeeds

### PR DESCRIPTION
**Pull request by HonoreDB**
BIT and VARBIT are bitfield types. For efficiency, we're encoding them as
 arrays of 64-bit integers.Since our encoding is big-endian,
the last value may have many trailing zeroes. For this reason,
the first value of each array gives the number of bits that are
used in the last value. For example, if the bitfield is 65 bits long,
this will be encoded into an array that looks like

[1, <first 64 bits>, either 0 or the 64-bit integer with only the first bit set].

This is admittedly a little wasteful in the case of BIT( number less than 64), but it seems best to be
consistent.

Release note (bug fix): avro changefeeds now support BIT and VARBIT columns

### CODE REVIEW

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-Mb8LB3c-7es_RM7ZO8J:b-2f36mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Could (or should) the allocs be memoized?  You could allocate [2]int64 for common case once; allocate if you need more than 1 word.  Or, you can allocate slice w/ 2 elements and reallocate it as needed?
</blockquote></details>

I hadn't thought of this and it seems worthwhile, but in implementing it I've run into a the same problem as with reusing the union key map: you can have an array column of type `VARBIT[]` or `BIT(65)[]`, and in that case goavro expects the full native array to be passed in as one argument, so I can't reuse any memory between elements of an array. Solvable, but probably not elegantly--I'll need to make the item schema know it's an item schema and where it is in the array.

It's annoying that I need to create this array in the first place; I'm exactly duplicating an existing array of unsigned integers just so I can tell the type system that no, it's okay that goavro expects signed integers. It's all getting encoded into binary anyway! Tempting to fork goavro to accept unsigned.

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-Mb8Yw--3FnGD4DrldwW:b-5jk54h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I hadn't thought of this and it seems worthwhile, but in implementing it I've run into a the same problem as with reusing the union key map: you can have an array column of type `VARBIT[]` or `BIT(65)[]`, and in that case goavro expects the full native array to be passed in as one argument, so I can't reuse any memory between elements of an array. Solvable, but probably not elegantly--I'll need to make the item schema know it's an item schema and where it is in the array.

It's annoying that I need to create this array in the first place; I'm exactly duplicating an existing array of unsigned integers just so I can tell the type system that no, it's okay that goavro expects signed integers. It's all getting encoded into binary anyway! Tempting to fork goavro to accept unsigned.
</blockquote></details>

Addressed, although let me know if you think this actually makes sense. For both groups of allocs, I'm now creating a pool and growing it through append() as needed, so that we only get a new alloc when an array is larger than this schema has ever seen.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-Mb8axSJ1cnGdKvPteCp:b-9etj2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Addressed, although let me know if you think this actually makes sense. For both groups of allocs, I'm now creating a pool and growing it through append() as needed, so that we only get a new alloc when an array is larger than this schema has ever seen.
</blockquote></details>

I like that we no longer have `reuseMap` parameter.  But we got quite a bit more complexity -- and as you say, it's not clear if it's actually worth it.

I want to revisit the way array encoding was done.
Basically, I think we can change the encoder function passed to `setNullable` to have a different signature:

```
encoder func(datum tree.Datum, memo interface{}) (interface{}, error),
```

The memo value passed in will be the current value stored in the map[string]interface{}.
The function may, or may not use it -- and array encoder of course should use it.
See https://github.com/cockroachdb/cockroach/pull/65970/files where it appears that all avro tests pass, and the array encoder benchmark seems to be ~30% faster:

```
Before:
BenchmarkEncodeIntArray-16    	 1118847	      1368 ns/op	     873 B/op	      13 allocs/op

After:
BenchmarkEncodeIntArray-16    	 2080842	      1099 ns/op	     177 B/op	       9 allocs/op
```

Do you think something like this would work for bit arrays too?

**Comment by HonoreDB** [comment]
___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-MbDVVp90dx0V57UfKAJ:b9ltkcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I like that we no longer have `reuseMap` parameter.  But we got quite a bit more complexity -- and as you say, it's not clear if it's actually worth it.

I want to revisit the way array encoding was done.
Basically, I think we can change the encoder function passed to `setNullable` to have a different signature:

```
encoder func(datum tree.Datum, memo interface{}) (interface{}, error),
```

The memo value passed in will be the current value stored in the map[string]interface{}.
The function may, or may not use it -- and array encoder of course should use it.
See https://github.com/cockroachdb/cockroach/pull/65970/files where it appears that all avro tests pass, and the array encoder benchmark seems to be ~30% faster:

```
Before:
BenchmarkEncodeIntArray-16    	 1118847	      1368 ns/op	     873 B/op	      13 allocs/op

After:
BenchmarkEncodeIntArray-16    	 2080842	      1099 ns/op	     177 B/op	       9 allocs/op
```

Do you think something like this would work for bit arrays too?
</blockquote></details>

I think these are equivalent optimizations? The benchmark for IntArray on my branch fluctuates from 8 to 13 allocs--is there a way to set N higher so we can get robust comparisons? I've just been running it from inside vscode.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-MbDeQ8R3tLBuUq9Veom:b-k885yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I think these are equivalent optimizations? The benchmark for IntArray on my branch fluctuates from 8 to 13 allocs--is there a way to set N higher so we can get robust comparisons? I've just been running it from inside vscode.
</blockquote></details>

Not sure... In your case we're changing avroSchemaField struct for all types .  We also carry indexes as part of main type -- even though those are applicable only to the arrays.

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-MbG9-7ACWiatiE2BDa1:b-9f53o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
> is there a way to set N higher so we can get robust comparisons?

It looks like there is a -benchtime flag: https://golang.org/cmd/go/#hdr-Testing_flags

Looks like something like this:

```
 make bench GOTESTFLAGS="-benchtime 30s" -v" PKG=./pkg/ccl/changefeedccl BENCHES= BenchmarkEncodeIntArray
```

or this:

```
 make bench GOTESTFLAGS="-benchtime 4000000x " -v" PKG=./pkg/ccl/changefeedccl BENCHES= BenchmarkEncodeIntArray
```

might work.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-MbGbG0V6XTA1Ti9f7rE:b-k0rbz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

> is there a way to set N higher so we can get robust comparisons?

It looks like there is a -benchtime flag: https://golang.org/cmd/go/#hdr-Testing_flags

Looks like something like this:

```
 make bench GOTESTFLAGS="-benchtime 30s" -v" PKG=./pkg/ccl/changefeedccl BENCHES= BenchmarkEncodeIntArray
```

or this:

```
 make bench GOTESTFLAGS="-benchtime 4000000x " -v" PKG=./pkg/ccl/changefeedccl BENCHES= BenchmarkEncodeIntArray
```

might work.
</blockquote></details>

@honoredb to be clear: it's less about performance and more about code org.
You mentioned that we implemented varbit w/out memoization because of array - so, arrays seem to be the "culprit" here.  I tried to leverage the fact that we store interface{} in the native field to move all of the array specific complexity logic internally.  I like this better than changing main struct and carrying indexes, etc because all of these changes are applicable only to arrays.
If you can find other and/or better ways to do this -- then go for it.  If you need to change setNullable to be something else entirely -- go for it if it improves the code.

**Comment by HonoreDB** [comment]
___
*[pkg/ccl/changefeedccl/avro.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MapIzg_hqf0tXVlUlaZ-r1-246:-MbMfMF22V-hPCQiGWbV:b-3i04b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce8072ef47946b377c0147d6497132c163312562/pkg/ccl/changefeedccl/avro.go#L246)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

@honoredb to be clear: it's less about performance and more about code org.
You mentioned that we implemented varbit w/out memoization because of array - so, arrays seem to be the "culprit" here.  I tried to leverage the fact that we store interface{} in the native field to move all of the array specific complexity logic internally.  I like this better than changing main struct and carrying indexes, etc because all of these changes are applicable only to arrays.
If you can find other and/or better ways to do this -- then go for it.  If you need to change setNullable to be something else entirely -- go for it if it improves the code.
</blockquote></details>

Gotcha. I've pulled in your solution, then expanded it a bit so that we still get the savings from mine--we now pass in a memo even in the array item case, by tracking the index in the array encoder function instead of on the item schema.

**Review by miretskiy — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[pkg/ccl/changefeedccl/avro.go, line 592 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MbN-_BACUSF8ua-ewG5:-MbN-_BACUSF8ua-ewG6:be7dbh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/23eb16c8064ad216833a9b5c8ef863e3acdf8061/pkg/ccl/changefeedccl/avro.go#L592)):*
> ```Go
> 							avroArr = append(avroArr, encMap)
> 						}
> 					}
> ```

Even though I contributed this code -- I don't trust myself.  Perhaps it's too complex for its own good.
I would either expand testing -- (a lot) -- to be confident in this code, or perhaps simplify it by doing
what you've done for bit array -- namely, if memoize value is not large enough -- just reallocate it.
wdyt?

**Comment by HonoreDB** [comment]
___
*[pkg/ccl/changefeedccl/avro.go, line 592 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MbN-_BACUSF8ua-ewG5:-MbN3Lf3COzus-35zBbI:b-8c7e4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/23eb16c8064ad216833a9b5c8ef863e3acdf8061/pkg/ccl/changefeedccl/avro.go#L592)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Even though I contributed this code -- I don't trust myself.  Perhaps it's too complex for its own good.
I would either expand testing -- (a lot) -- to be confident in this code, or perhaps simplify it by doing
what you've done for bit array -- namely, if memoize value is not large enough -- just reallocate it.
wdyt?
</blockquote></details>

Would you trust it enough if I added a few more arrays to the TestAvroArray list, to make sure we see all the cases where arrays grow and shrink by multiple elements at a time and can have null values at the edges? I think we can get away with the new allocs in the bit array because it's going to be really rare for array size to grow there, but it might be pretty common in the regular array case.

**Comment by miretskiy** [comment]
Yeah. Either expanded test coverage or simpler code

**Comment by HonoreDB** [comment]
___
*[pkg/ccl/changefeedccl/avro.go, line 592 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65869#-MbN-_BACUSF8ua-ewG5:-MbNXI_w9ehCI0b-y-Fq:b-n8btgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/23eb16c8064ad216833a9b5c8ef863e3acdf8061/pkg/ccl/changefeedccl/avro.go#L592)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Would you trust it enough if I added a few more arrays to the TestAvroArray list, to make sure we see all the cases where arrays grow and shrink by multiple elements at a time and can have null values at the edges? I think we can get away with the new allocs in the bit array because it's going to be really rare for array size to grow there, but it might be pretty common in the regular array case.
</blockquote></details>

Added some more arrays in the test. Also deleted a little code: I couldn't find a way to actually hit the code path that tests for `i` being between the memoized slice's length and capacity, and I don't think we actually need it because `append` will take advantage of excess capacity when it's there and grow the slice in place.

**Review by miretskiy — APPROVED** [review]
> Reviewed 2 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

**Comment by HonoreDB** [comment]
bors r=[miretskiy]
